hkmc2/shared/src/test/mlscript-compile/Rendering.mls

```
@@ -35,27 +35,350 @@ fun interleave(sep)(...args) =
     if i < len do set res.[idx + 1] = sep
   res
 
-fun render(arg) = if
-  arg is
-    undefined then "undefined"
-    null      then "null"
-    Array     then fold(+)("[", ...interleave(", ")(...map(render)(...arg)), "]")
-    Str       then JSON.stringify(arg)
-    Set       then fold(+)("Set{", ...interleave(", ")(...map(render)(...arg)), "}")
-    Map       then fold(+)("Map{", ...interleave(", ")(...map(render)(...arg)), "}")
-    Function  and
-      let p = Object.getOwnPropertyDescriptor(arg, "prototype")
-      (p is Object and p.("writable")) || (p is undefined) then
-        "[function" + (if arg.name is
-          ""  then ""
-          nme then " " + nme
-        ) + "]"
-    Object then
-      if arg.constructor.name is "Object"
-      then
-        let es = Object.entries(arg)
-        fold(+)("{", ...interleave(", ")(...map(case [k, v] then k + ": " + render(v))(...es)), "}")
-      else String(arg)
-  let ts = arg.("toString") // not accessing as `arg.toString` to avoid the sanity check
-  ts is undefined then "[" + typeof(arg) + "]"
-  else ts.call(arg)
+// * Apply `f` to `xs` and build a string from non-null results using the given
+// * separator and starting/ending symbols. If there is no non-null results,
+// * return `empty` instead.
+fun mkString(xs, f, start, sep, end, empty) =
+  let
+    it = xs.(Symbol.iterator)()
+    res = it.next()
+    buffer = null
+  // Consume the iterator until we find the first non-null value.
+  while res.done is false and buffer is null do
+    if f(res.value) is
+      Str as str do set buffer = start + str
```

You are silently ignoring all non-string values? 🤔 

Better throw an exception, no?

---

This function (`mkString`) is not used and will be removed.